src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorUpdaterStep.cs

```diff
@@ -0,0 +1,148 @@
+    /// analyze RazorCodeDocument objects and make necessary updates
+    /// to cshtml files.
+    /// </summary>
+    public class RazorUpdaterStep : UpgradeStep
+    {
+        private RazorProjectEngine? _razorEngine;
+
+        public override string Title => "Update Razor files";
+
+        public override string Description => "Update Razor files using registered Razor updaters";
+
+        public override string Id => WellKnownStepIds.RazorUpdaterStepId;
+
+        public override IEnumerable<string> DependsOn { get; } = new[]
+        {
+            // Project should be backed up before changing source
+            WellKnownStepIds.BackupStepId,
+
+            // Template files should be added prior to changing source (since some code fixers will change added templates)
+            WellKnownStepIds.TemplateInserterStepId,
+
+            // Project should have correct TFM
+            WellKnownStepIds.SetTFMStepId,
+        };
+
+        public override IEnumerable<string> DependencyOf { get; } = new[]
+        {
+            WellKnownStepIds.NextProjectStepId,
+        };
+
+        private ImmutableArray<RazorCodeDocument>? _razorDocuments;
+
+        public ImmutableArray<RazorCodeDocument> RazorDocuments =>
+            _razorDocuments ?? throw new InvalidOperationException("Razor documents are not available until the updater step is initialized or until ProcessRazorDocuments has been called.");
+
+        public RazorUpdaterStep(IEnumerable<IUpdater<RazorCodeDocument>> razorUpdaters, ILogger<RazorUpdaterStep> logger)
+            : base(logger)
+        {
+            if (razorUpdaters is null)
+            {
+                throw new ArgumentNullException(nameof(razorUpdaters));
+            }
+
+            if (logger is null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            // Add a sub-step for each Razor updater
+            SubSteps = new List<UpgradeStep>(razorUpdaters.Select(updater => new RazorUpdaterSubStep(this, updater, logger)));
+        }
+
+        /// <summary>
+        /// Determines whether the RazorUpdaterStep applies to a given context. This doesn't
+        /// consider whether there's actually any work to be done by Razor updaters. Rather, it
+        /// determines whether it makes sense to even initialize the Razor updaters. So, it
+        /// will return false if there are no cshtml files in the current project or if there
+        /// are no Razor updaters available.
+        /// </summary>
+        /// <param name="context">The context to evaluate.</param>
+        /// <param name="token">A token that can be used to cancel execution.</param>
+        /// <returns>True if the Razor updater step might apply, false otherwise.</returns>
+        protected override Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token) =>
+            Task.FromResult(context?.CurrentProject is not null
+            && SubSteps.Any()
```

Comment:
Can we ask the Razor substeps if they apply (similar to how we ask IUpdater<ConfigFile> substeps if they apply)

Reply:
Good idea. I'll make that update.